src/storage/src/source/reclock.rs

```diff
@@ -47,6 +47,91 @@ pub struct ReclockFollower {
     inner: Rc<RefCell<ReclockFollowerInner>>,
 }
 
+/// Packs a binding into a Row.
+///
+/// A binding of None partition is encoded as a single datum containing the offset.
+///
+/// A binding of a Kafka partition is encoded as the partition datum followed by the offset datum.
+fn pack_binding(pid: PartitionId, offset: MzOffset) -> SourceData {
+    let mut row = Row::with_capacity(2);
+    let mut packer = row.packer();
+    match pid {
+        PartitionId::None => {}
+        PartitionId::Kafka(pid) => packer.push(Datum::Int32(pid)),
+    }
+    packer.push(Datum::UInt64(offset.offset));
+    SourceData(Ok(row))
+}
+
+/// Unpacks a binding from a Row
+/// See documentation of [pack_binding] for the encoded format
+fn unpack_binding(data: SourceData) -> (PartitionId, MzOffset) {
+    let row = data.0.expect("invalid binding");
+    let mut datums = row.iter();
+    let (pid, offset_datum) = match (datums.next(), datums.next()) {
+        (Some(Datum::Int32(pid)), Some(offset_datum)) => (PartitionId::Kafka(pid), offset_datum),
+        (Some(offset_datum), None) => (PartitionId::None, offset_datum),
+        _ => panic!("invalid binding"),
+    };
+
+    let offset = match offset_datum {
+        Datum::UInt64(offset) => MzOffset::from(offset),
+        _ => panic!("invalid binding"),
+    };
+    (pid, offset)
+}
+
+/// Drains the provided vector of updates containing insertions and retractions of full offset
+/// statements and differentiates them into updates containing the change in offset in the diff
+/// field.
+fn differentiate(
+    updates: &mut Vec<((PartitionId, MzOffset), Timestamp, Diff)>,
+) -> impl Iterator<Item = (PartitionId, Timestamp, MzOffset)> + '_ {
+    consolidation::consolidate_updates(updates);
+    let mut prev_update = None;
+    updates
+        .drain(..)
+        .filter_map(move |((pid, offset), ts, diff)| match diff {
+            -1 => {
+                prev_update = Some((pid, offset, ts));
+                None
+            }
+            1 => {
+                let prev_offset = match prev_update.take() {
+                    Some((prev_pid, prev_offset, prev_ts)) => {
+                        assert_eq!(prev_pid, pid, "invalid bindings");
+                        assert_eq!(prev_ts, ts, "invalid bindings");
+                        prev_offset
+                    }
+                    None => MzOffset::from(0),
+                };
+                offset
+                    .checked_sub(prev_offset)
+                    .map(move |diff| (pid, ts, diff))
+            }
```

Comment:
might be nice to add a comment like "we effectively turn pairs of `source_upper` retractions+insertions (placed next to each other in the stream by `integrate` into mzoffset diffs"

Reply:
added